### ISSUE TITLE
Bogus `-Wtype-safety` warnings in MPICH when using const buffers

### ISSUE THREAD

**Issue by msimberg** [llvm/llvm-project#62512]
I'm "cross-posting" an issue from https://github.com/pmodels/mpich/issues/6496 because I'm not sure what the most appropriate place for this is:

MPICH uses attributes from clang that are supposed to help catch type errors where the type of a buffer is not the same as the `MPI_Datatype` passed in to a function (https://clang.llvm.org/docs/AttributeReference.html#type-safety-checking). In functions that take a `const void*` buffer clang reports bogus (as far as I can tell) warnings where the data type is correct, but the constness is different. E.g.
```
const int x;
MPI_Send(&x, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
```
reports `warning: argument type 'const int *' doesn't match specified 'MPI' type tag that requires 'int *' [-Wtype-safety]`. The only difference is `const` and this call should not trigger a warning.

Workarounds include:
- Disabling `-Wtype-safety` either locally or globally. This is not satisfying because it completely turns off checking and mismatches are not at all reported. E.g.
```
int x;
MPI_Send(&x, 1, MPI_CHAR, 1, 0, MPI_COMM_WORLD);
```
would not produce a warning.
- `const_cast` because we know `MPI_Send` anyway takes a `const` pointer but this is verbose and requires repeating the type:
```
const int x;
MPI_Send(const_cast<int*>(&x), 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
```
- Making a const `MPI_INT`. This is again verbose and requires repeating the type:
```
auto const MPI_CONST_INT = MPI_INT;
const int x;
MPI_Send(&x, 1, MPI_CONST_INT, 1, 0, MPI_COMM_WORLD);
```
- Disable the attribute when configuring MPICH. This is not always an option.

Here's a short reproducer on godbolt (without MPICH, but using the attributes): https://godbolt.org/z/M8P1oqz44.

Given that the MPICH developers think that this is a clang issue https://github.com/pmodels/mpich/issues/6496#issuecomment-1531691516: is it? That is:
- Is this working as intended? The warning does make sense in isolation because `int*` and `const int*` are not the same type. On the other hand, it does not make sense in the context of MPICH where `MPI_INT` is there only to signal the size of the type, and not the constness.
- Is there currently a way for MPICH to use the attributes while ignoring constness? If yes, what would that be?
- If there's no way currently to treat `int*` and `const int*` as the same for the purposes of the warnings, would it make sense to add an option to the attribute to support that?

**Comment by prj-** [comment]
@EugeneZelenko, the issue is still present with `clang version 20.0.0 (++20240825100934+5b41eb3a6dcd-1~exp1~20240825221045.2294)`, is there any hope that it gets fixed?
```
$ clang-20 dump.c 
dump.c:15:14: warning: argument type 'const int *' doesn't match specified 'mpi' type tag that requires 'int *' [-Wtype-safety]
   15 |     MPI_Send(&x, 1, MPI_INT);
      |              ^~     ~~~~~~~
1 warning generated.
```

**Comment by prj-** [comment]
Or maybe @AaronBallman could help us move forward with this?

**Comment by AaronBallman** [comment]
Hmmm, I think you can [ab]use [`layout_compatible`](https://clang.llvm.org/docs/AttributeReference.html#type-tag-for-datatype) to solve this, right?

https://godbolt.org/z/r7M8Kx7n9

CC @gribozavr for opinions on whether the current behavior is or isn't a bug.

**Comment by prj-** [comment]
What do you say @hzhou and @raffenet? MPICH is currently triggering tons of warning on macOS Sequoia Beta 8 because of this.

**Comment by prj-** [comment]
It appears the `layout_compatible` clause was just recently added (I still get warnings when not using `trunk`), so we'll be stuck with those warnings for a while if we go down that road :(
